### ISSUE TITLE
Be able to specify to pixel dimensions for maps which aren't attached to the DOM

### ISSUE THREAD

**Issue by edg2s** [Leaflet/Leaflet#4200]
Hi, I work for the WIkimedia Foundation and I'm currently helping integrating OSM maps with Wikipedia.

One issue we have is that the maps need to be re-rendered when a 'content changed' event fires on the page. For performance this event fires when the content is ready to be attached, and not after (to avoid re-paints). Leaflet only measures dimensions using _container.clientWidth, which at this point is always 0, even though the caller knows the final dimensions and has already set the style attributes for them.

Ideally there would be an API to allow users to efficiently prepare maps before attaching them to the DOM, something like setSize(L.Point).

**Comment by IvanSanchez** [comment]
Hi, @edg2s ! Happy to know folks at Wikipedia are using Leaflet :-)

I've read the phabricator issue, and one thing is not clear to me: do you use Leaflet 0.7.x or 1.0.0-beta?

Anyway, you should be able to hack your way around this by redefining the `L.Map.getSize()` method, like:

```js
var map = L.map(detachedContainer, mapOptions);

map.getSize = function() { return L.point(400,300); };

map.addLayer(...);
map.setView(...);
```

It would be possible to add another map option to specify a initial map size (valid until the first call to `L.Map.invalidateSize()` happens), change `L.Map.initialize()` to use that option to set `this._size` (to the value of the option) and `this._sizeChanged` (to `false`) and test, test, test.

**Comment by edg2s** [comment]
Yeah we've hacked around it by setting the _size property directly. I've leaving getSize as it is possible to change the maps size later with the editor. It looks like we're currently using 0.7.7.

If you're setting initial size you'd want to make sure getSize doesn't invalidate if the element is hidden, but the quickest way to check that is clientWidth=clientHeight=0 which isn't perfect.

CCing @nyurik

**Comment by nyurik** [comment]
@IvanSanchez, at first we I used Leaflet directly, but later we needed [Simple styling](https://github.com/mapbox/simplestyle-spec/tree/master/1.1.0), so we switched to [mapbox.js lib](https://github.com/mapbox/mapbox.js) (which uses Leaflet internally). At some point I hope we can start using webGL client-side rendering as well, since our maps are vector based, as well as integrate leaflet editor for geojson/topojson overlay editing.

See [demo here](http://vem3.wmflabs.org/wiki/Main_Page).  Try "edit" and insert/modify a map. Thanks for your help!

**Comment by jasongrout** [comment]
I'd really like to set the size manually too. I'm using leaflet in the JupyterLab project, based on the PhosphorJS web library. In Phosphor, all things are layed out with absolute sizes, which I get in resize messages. In short, this means that when I create the map or when it needs to be resized, I know the exact dimensions, and I'd like to avoid the DOM read.

**Comment by jasongrout** [comment]
Would you accept a PR making a setSize() function, and updating `invalidateSize` to take a second optional argument, which would be a Point object representing the new size? If the optional argument was passed in, invalidateSize would call setSize, rather than getSize?

**Comment by SylvainCorlay** [comment]
:+1: on being able to set the size manually.

**Comment by SylvainCorlay** [comment]
We would use it for the ipyleaflet module, which will soon be using phosphor too.
